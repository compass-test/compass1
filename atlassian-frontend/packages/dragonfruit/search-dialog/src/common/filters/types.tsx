import { Products } from '../product-context';

export enum FilterOptionSource {
  /**
   * This means the result came from collab graph
   */
  COLLABORATION_GRAPH = 'COLLABORATION_GRAPH',
  /**
   * This means the result came from smart user picker
   */
  SMART_USER_PICKER = 'SMART_USER_PICKER',
  /**
   * This means the result came from a recents (bootstrap) query
   */
  RECENT = 'RECENT',
  /**
   * This means the result came from an explicit search
   */
  SEARCH = 'SEARCH',
  /**
   * This means the result is derived from the current user
   */
  CURRENT_USER = 'CURRENT_USER',
  /**
   * This means the result came from an external source passed into the component
   */
  EXTERNAL = 'EXTERNAL',
  /**
   * This means the result is static without any user-generation
   */
  STATIC = 'STATIC',
}

export interface FilterOption {
  id: string;
  isChecked: boolean;
  isVisible: boolean;
  filterSource: FilterOptionSource;
}

export interface SiteFilterOption extends FilterOption {
  product: Products;
  avatarUrl: string;
  cloudId: string;
  siteName: string;
  siteUrl: string;
}
