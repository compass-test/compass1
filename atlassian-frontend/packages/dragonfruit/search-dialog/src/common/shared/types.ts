export interface ResultState<T> {
  isLoading: boolean;
  results: T | null;
}

export enum ActionSubjectId {
  preQuerySearchResults = 'preQuerySearchResults',
  postQuerySearchResults = 'postQuerySearchResults',
  cachedResults = 'cachedResults',
}
