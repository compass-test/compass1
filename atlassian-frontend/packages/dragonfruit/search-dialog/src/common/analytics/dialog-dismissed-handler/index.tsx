export { DialogDismissedHandler } from './dialog-dismissed-handler';
