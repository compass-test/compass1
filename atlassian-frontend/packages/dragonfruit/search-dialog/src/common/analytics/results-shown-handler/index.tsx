export { default as SearchResultsShownHandler } from './results-shown-handler';
