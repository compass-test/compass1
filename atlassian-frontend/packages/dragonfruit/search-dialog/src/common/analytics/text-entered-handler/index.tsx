export { default as TextEnteredHandler } from './text-entered-handler';
