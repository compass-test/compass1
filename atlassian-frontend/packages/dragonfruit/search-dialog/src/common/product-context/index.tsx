export {
  Products,
  useActiveProduct,
  usePrimaryProduct,
  useProducts,
  ProductProvider,
} from './product-context';
