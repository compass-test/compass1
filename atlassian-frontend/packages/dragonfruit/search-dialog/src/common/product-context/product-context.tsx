import React, { useContext, FunctionComponent } from 'react';

export enum Products {
  confluence = 'CONFLUENCE',
  compass = 'COMPASS',
  other = 'OTHER',
}

interface Props {
  products: Array<Products>;
  activeIndex: number;
  primaryProduct?: string; // To be populated if this component is used in anything other than: enum Products
}

const primaryProductIndex = 0;

const ProductContext = React.createContext<Props>({
  products: [],
  activeIndex: primaryProductIndex,
  primaryProduct: '',
});

interface ProductProviderProps {
  products: Array<Products>;
  activeIndex?: number;
  primaryProduct?: string;
}

export const ProductProvider: FunctionComponent<ProductProviderProps> = ({
  children,
  products,
  activeIndex = 0,
  primaryProduct = '',
}) => (
  <ProductContext.Provider
    value={{
      products,
      activeIndex: Math.min(products.length - 1, Math.max(0, activeIndex)),
      primaryProduct,
    }}
  >
    {children}
  </ProductContext.Provider>
);

export const useProducts = () => useContext(ProductContext).products;

// Returns primaryProduct if defined else the first product is considered the primary
export const usePrimaryProduct = () => {
  const { products, primaryProduct } = useContext(ProductContext);

  return primaryProduct || products[primaryProductIndex];
};

export const useActiveProduct = () => {
  const context = useContext(ProductContext);
  return context.products[context.activeIndex];
};
