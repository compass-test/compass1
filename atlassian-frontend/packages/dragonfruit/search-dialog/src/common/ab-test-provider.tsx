import React, {
  useContext,
  FunctionComponent,
  useState,
  useEffect,
} from 'react';
import { ABTest, SharedClient } from './clients';
import { SearchDialogAnalyticsContext, useAnalytics } from './analytics';
import { onExperimentLoaded } from './analytics/events';
import { ProductSearchInputSkeleton } from './product-search-input-skeleton';

export const DEFAULT_AB_TEST: ABTest = {
  abTestId: 'default',
  controlId: 'default',
  experimentId: 'nav-v3',
};

const MISSING_PROVIDER_ERROR = new Error(
  'Experimentation data has not been initialised, please ensure that there is an ABTestContext.Provider available',
);
interface Props {
  searchClient: SharedClient;
  isMultiProduct?: boolean;
}

interface State {
  abTest: ABTest | null;
}

export const ABTestContext = React.createContext<State>({
  abTest: null,
});

export const ABTestProvider: FunctionComponent<Props> = ({
  searchClient,
  isMultiProduct,
  children,
}) => {
  const [abTest, setAbTest] = useState<ABTest | null>(null);
  const { fireAnalyticsEvent } = useAnalytics();

  useEffect(() => {
    const now = Date.now();

    searchClient
      .getAbTestData()
      .then((abTestResponse) => {
        const duration = Date.now() - now;
        const abTestWithDefault = abTestResponse || DEFAULT_AB_TEST;
        setAbTest(abTestWithDefault);
        fireAnalyticsEvent(
          onExperimentLoaded(abTestWithDefault, duration, !!isMultiProduct),
        );
      })
      .catch((e) => {
        setAbTest(DEFAULT_AB_TEST);
      });
  }, [searchClient, fireAnalyticsEvent, isMultiProduct]);

  if (!abTest) {
    return <ProductSearchInputSkeleton />;
  }

  return (
    <ABTestContext.Provider value={{ abTest }}>
      <SearchDialogAnalyticsContext
        analyticContext={{ abTest }}
        nonPrivacySafeAnalyticContext={{}}
      >
        {children}
      </SearchDialogAnalyticsContext>
    </ABTestContext.Provider>
  );
};

export const useABTest = () => {
  const { abTest } = useContext(ABTestContext);

  if (!abTest) {
    throw MISSING_PROVIDER_ERROR;
  }

  return abTest;
};
