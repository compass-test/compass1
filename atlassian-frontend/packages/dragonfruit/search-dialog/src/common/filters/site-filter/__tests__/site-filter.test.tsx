import React from 'react';
import SiteFilter, { SiteFilterItem, SiteFilterProps } from '../site-filter';
import { mountWithIntl } from '../../../../__tests__/__fixtures__/intl-test-helpers';
import { FilterShowMore } from '@atlassian/search-dialog';
import { FilterOptionSource, SiteFilterOption } from '../../types';
import { useAnalytics, onFilterSelect } from '../../../../common/analytics';
import { createSiteFilters } from '../../../../__tests__/__fixtures__/mock-filters';
import { MAX_SELECTED_SITES } from '../site-filter';
import { Products } from '../../../product-context';

jest.mock('../../../../common/analytics', () => {
  const fireAnalyticsEvent = jest.fn();

  return {
    ...(jest.requireActual('../../../../common/analytics') as Object),
    useAnalytics: () => ({
      fireAnalyticsEvent,
    }),
  };
});

const sites: SiteFilterOption[] = createSiteFilters(2, {
  checkedNumber: 1,
  visibleNumber: 1,
});

const onChange = jest.fn();

const defaultProps: SiteFilterProps = {
  availableFilters: sites,
  onChange,
};

describe('<SiteFilter />', () => {
  beforeEach(jest.clearAllMocks);

  it('renders show more when at least one site is not visible', () => {
    const wrapper = mountWithIntl(<SiteFilter {...defaultProps} />);

    expect(wrapper.find(FilterShowMore)).toHaveLength(1);
  });

  it('does not render show more when all sites are visible', () => {
    const allVisibleSites = defaultProps.availableFilters.map((s) => ({
      ...s,
      isVisible: true,
    }));

    const wrapper = mountWithIntl(
      <SiteFilter {...defaultProps} availableFilters={allVisibleSites} />,
    );

    expect(wrapper.find(FilterShowMore)).toHaveLength(0);
  });

  it('disables show more when the max number of sites are selected (with more available to become visible)', () => {
    const maxVisibleSites = createSiteFilters(MAX_SELECTED_SITES + 1, {
      visibleNumber: MAX_SELECTED_SITES,
      checkedNumber: MAX_SELECTED_SITES,
    });
    const wrapper = mountWithIntl(
      <SiteFilter {...defaultProps} availableFilters={maxVisibleSites} />,
    );

    expect(wrapper.find(FilterShowMore).prop('isDisabled')).toEqual(true);
  });

  it('disables unselected checkboxes once the max selected checkboxes is reached', () => {
    const sites = createSiteFilters(2 * MAX_SELECTED_SITES, {
      visibleNumber: 2 * MAX_SELECTED_SITES,
      checkedNumber: MAX_SELECTED_SITES,
    });
    const wrapper = mountWithIntl(
      <SiteFilter {...defaultProps} availableFilters={sites} />,
    );

    expect(
      wrapper.find(SiteFilterItem).filterWhere((f) => f.prop('isDisabled')),
    ).toHaveLength(MAX_SELECTED_SITES);
  });

  it('Add new option from dropdown adds new option checked and fires analytics', () => {
    const sites = defaultProps.availableFilters;
    const wrapper = mountWithIntl(
      <SiteFilter {...defaultProps} availableFilters={sites} />,
    );

    (wrapper.find(FilterShowMore).prop('addFilter') as any)({
      value: sites[1],
    });

    wrapper.update();

    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(sites[1].id, true);

    // Analytics
    expect(useAnalytics().fireAnalyticsEvent).toBeCalledTimes(1);
    expect(useAnalytics().fireAnalyticsEvent).toBeCalledWith(
      onFilterSelect({
        filterId: sites[1].id,
        filterSource: sites[1].filterSource,
        filterType: 'site',
      }),
    );
  });

  it('should pass through the sites which matches the entered query in Show More & excludes already visible option', async () => {
    const sites = [
      {
        avatarUrl: 'http://adrain.name',
        cloudId: '5c3f2368-3393-4e09-8492-fd03e808d9d8',
        filterSource: FilterOptionSource.EXTERNAL,
        id: '5c3f2368-3393-4e09-8492-fd03e808d9d8',
        isChecked: false,
        isVisible: false,
        product: Products.confluence,
        siteName: 'ac/dc', // query doesn't match site name
        siteUrl: 'https://google.com',
      },
    ];
    const wrapper = mountWithIntl(
      <SiteFilter {...defaultProps} availableFilters={sites} />,
    );

    const promise = (wrapper.find(FilterShowMore).prop('loadOptions') as any)(
      'test',
    );
    wrapper.update();
    await new Promise((resolve) => setTimeout(resolve, 350));

    promise.then((results: any) => {
      expect(results).toMatchSnapshot();
    });
  });
});
