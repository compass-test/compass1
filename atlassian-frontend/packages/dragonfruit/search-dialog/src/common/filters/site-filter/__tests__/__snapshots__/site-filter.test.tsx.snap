// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteFilter /> should pass through the sites which matches the entered query in Show More & excludes already visible option 1`] = `Array []`;
